### PR TITLE
feat: [PL-31520]: Dynamic favicon support based on pipeline execution status

### PR DESCRIPTION
**Pull request by samarth-sharma31**
### Summary

Adds support to dynamically update the favicon based on pipeline execution status

#### Screenshots

Screenshot 2023-03-14 at 2 40 54 PM

https://user-images.githubusercontent.com/96409598/224975292-066d5421-fa5c-43b5-9862-e13261a1c545.mov

https://user-images.githubusercontent.com/96409598/224975365-257d032b-be15-4b1b-bce0-356ada434782.mov

#### PR Checklist

<details>
<summary>Please check if your PR fulfils the following requirements</summary>

- [ ] Tests for the changes have been added. Ideally, include a test that fails without this PR but passes with it.
- [ ] Docs have been [added/updated](https://harness.atlassian.net/jira/software/c/projects/DOC/boards/40).
</details>

<details>
<summary>Use the following comments to re-trigger PR Checks</summary>

- Jest: `retrigger jest`
- Prettier: `retrigger prettier`
- Type Check: `retrigger typecheck`
- ESLint: `retrigger eslint`
- Sonar: `retrigger sonar`
- Standards: `retrigger standards`
- Build: `retrigger build`
- Title Check: `retrigger titlecheck`
- Feature Name Check: `trigger featurenamecheck`
- Coverage: `retrigger coverage`
- Rebase: `trigger rebase`
- Cypress Rest: `retrigger cypress-rest`
- Cypress CD: `retrigger cypress-cd`
- Cypress Pipeline: `retrigger cypress-pipeline`
- Cypress CV: `retrigger cypress-cv`
- Fix Prettier: `fix prettier`
</details>

### CODE REVIEW

File: src/modules/70-pipeline/hooks/useDocumentFavicon.ts [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+  }
+
+  if (isExecutionRunningLike(pipelineStatus)) {
+    return EXECUTION_STATUS_VS_FAVICON_INFO_MAP.RUNNING[favIconSourceType]
+  }
+
+  return DEFAULT_FAVICON_INFO[favIconSourceType]
+}
+
+export function useDocumentFavicon(pipelineStatus: PipelineExecutionSummary['status']) {
```
Comment (abhinavrastogi-harness) [inline src/modules/70-pipeline/hooks/useDocumentFavicon.ts:103]: let's move this hook to `10-common`.  other modules may also need to change the favicon in the future.
we can keep only the `getFavIconInfoFromStatus` util in `70-pipeline`.

File: src/modules/70-pipeline/hooks/useDocumentFavicon.ts [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+  const { selectedProject } = useAppStore()
+  // Favicons to be loaded statically instead of CDN for local and ONPREM env
+  // ref -> https://github.com/harness/harness-core-ui/pull/14420
+  const favIconSourceType = __DEV__ || isOnPrem() ? FavIconSourceTypes.STATIC : FavIconSourceTypes.CDN
```
Comment (abhinavrastogi-harness) [inline src/modules/70-pipeline/hooks/useDocumentFavicon.ts:107]: check `window.HARNESS_ENABLE_CDN` before setting CDN url to favicon.

File: src/modules/70-pipeline/hooks/useDocumentFavicon.ts [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    if (linkElementIco) {
+      linkElementIco.href = getFavIconHrefPath(favIconInfo.ICO)
+    }
+    const linkElementPng = document.getElementById('favicon-png') as HTMLLinkElement
+    if (linkElementPng) {
+      linkElementPng.href = getFavIconHrefPath(favIconInfo.PNG)
+    }
+  }
+
+  useDeepCompareEffect(() => {
```
Comment (abhinavrastogi-harness) [inline src/modules/70-pipeline/hooks/useDocumentFavicon.ts:121]: let's avoid `deepcompareEffect`.. instead use `useEffect` if none of the dependencies is a nested object.

File: src/modules/70-pipeline/hooks/useDocumentFavicon.ts [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    const favIconInfo = getFavIconInfoFromStatus(pipelineStatus, favIconSourceType)
+    updateFavicon(favIconInfo)
+
+    return () => {
+      // reset favicon on unmount
+      updateFavicon(DEFAULT_FAVICON_INFO[favIconSourceType])
+    }
+  }, [pipelineStatus, selectedProject, favIconSourceType])
```
Comment (abhinavrastogi-harness) [inline src/modules/70-pipeline/hooks/useDocumentFavicon.ts:129]: why is selectedProject a dependency here?

File: src/modules/70-pipeline/hooks/useDocumentFavicon.ts [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+export function useDocumentFavicon(pipelineStatus: PipelineExecutionSummary['status']) {
+  const { selectedProject } = useAppStore()
+  // Favicons to be loaded statically instead of CDN for local and ONPREM env
+  // ref -> https://github.com/harness/harness-core-ui/pull/14420
+  const favIconSourceType = __DEV__ || isOnPrem() ? FavIconSourceTypes.STATIC : FavIconSourceTypes.CDN
+
+  const updateFavicon = (favIconInfo: FavIconInfo): void => {
+    // logic to update favicon
+    const linkElementIco = document.getElementById('favicon-x-icon') as HTMLLinkElement
```
Comment (abhinavrastogi-harness) [inline src/modules/70-pipeline/hooks/useDocumentFavicon.ts:111]: can we cache the Link element using `useMemo` to prevent DOM lookup everytime?

**Review by abhinavrastogi-harness — CHANGES_REQUESTED** [review]

File: src/modules/70-pipeline/hooks/useDocumentFavicon.ts (hunk 5 above)
Comment (samarth-sharma31) [reply src/modules/70-pipeline/hooks/useDocumentFavicon.ts]: done!

File: src/modules/70-pipeline/hooks/useDocumentFavicon.ts (hunk 4 above)
Comment (samarth-sharma31) [reply src/modules/70-pipeline/hooks/useDocumentFavicon.ts]: yes, not required. Removed

File: src/modules/70-pipeline/hooks/useDocumentFavicon.ts (hunk 3 above)
Comment (samarth-sharma31) [reply src/modules/70-pipeline/hooks/useDocumentFavicon.ts]: done

File: src/modules/70-pipeline/hooks/useDocumentFavicon.ts (hunk 2 above)
Comment (samarth-sharma31) [reply src/modules/70-pipeline/hooks/useDocumentFavicon.ts]: done

File: src/modules/70-pipeline/hooks/useDocumentFavicon.ts (hunk 1 above)
Comment (samarth-sharma31) [reply src/modules/70-pipeline/hooks/useDocumentFavicon.ts]: done

File: src/modules/10-common/hooks/useDocumentFavicon.ts [hunk 6]
```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright 2021 Harness Inc. All rights reserved.
+ * Use of this source code is governed by the PolyForm Shield 1.0.0 license
+ * that can be found in the licenses directory at the root of this repository, also available at
+ * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
+ */
+
+import { useEffect, useMemo } from 'react'
+// eslint-disable-next-line no-restricted-imports
```
Comment (abhinavrastogi-harness) [inline src/modules/10-common/hooks/useDocumentFavicon.ts:9]: the eslint rule is in place for a reason. let's not bypass it 😅

File: src/modules/10-common/hooks/useDocumentFavicon.ts [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+import { getFavIconTypeFromPipelineStatus, SUPPORTED_FAVICON_STATUS_TYPES } from '@pipeline/utils/statusHelpers'
```
Comment (abhinavrastogi-harness) [inline src/modules/10-common/hooks/useDocumentFavicon.ts:10]: `useDocumentFavicon` should not be aware of pipelines. run the pipeline logic in the execution view, and then pass the final icon to this hook.

File: src/modules/10-common/hooks/useDocumentFavicon.ts [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+import type { PipelineExecutionSummary } from 'services/pipeline-ng'
```
Comment (abhinavrastogi-harness) [inline src/modules/10-common/hooks/useDocumentFavicon.ts:11]: same. common components should not depend on specific services. It'll make it impossible to move this component out of NGUI later, or for other microfrontends to reuse it.

File: src/modules/10-common/hooks/useDocumentFavicon.ts [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+    ICO: 'favicon.ico',
+    PNG: 'favicon.png'
+  },
+  CDN: {
+    ICO: 'https://static.harness.io/ng-static/images/favicon.ico',
+    PNG: 'https://static.harness.io/ng-static/images/favicon.png'
+  }
+}
+
+const EXECUTION_STATUS_VS_FAVICON_INFO_MAP = {
```
Comment (abhinavrastogi-harness) [inline src/modules/10-common/hooks/useDocumentFavicon.ts:35]: this map is specific to Pipelines.. it should not be in 10-common

File: src/modules/10-common/hooks/useDocumentFavicon.ts [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+    return `/${favIconIdentifier}`
+  }
+  if (isOnPrem()) {
+    return `/ng/static/${favIconIdentifier}`
+  }
+
+  return favIconIdentifier
+}
+
+const getFavIconInfoFromStatus = (
```
Comment (abhinavrastogi-harness) [inline src/modules/10-common/hooks/useDocumentFavicon.ts:99]: same. move to pipelines.

File: src/modules/10-common/hooks/useDocumentFavicon.ts [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+
+  return favIconType
+    ? EXECUTION_STATUS_VS_FAVICON_INFO_MAP[favIconType][favIconSourceType]
+    : DEFAULT_FAVICON_INFO[favIconSourceType]
+}
+
+// Favicons to be loaded statically instead of CDN for local and ONPREM env and also when CDN is not enabled
+// ref -> https://github.com/harness/harness-core-ui/pull/14420
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+const shouldUseStaticSource = () => __DEV__ || isOnPrem() || !(window as any).HARNESS_ENABLE_CDN
```
Comment (abhinavrastogi-harness) [inline src/modules/10-common/hooks/useDocumentFavicon.ts:113]: dont use `any`. instead declare this variable is available on Window in `global.d.ts`

File: src/modules/70-pipeline/utils/statusHelpers.ts [hunk 12]
```diff
@@ -205,3 +205,35 @@ export const isRetryPipelineAllowed = (status?: string): boolean => {
     isExecutionAborted(status)
   )
 }
+
+export enum SUPPORTED_FAVICON_STATUS_TYPES {
```
Comment (abhinavrastogi-harness) [inline src/modules/70-pipeline/utils/statusHelpers.ts:209]: why do we need a new enum? why not just use `PipelineExecutionSummary['status']` directly?

Comment (samarth-sharma31) [reply src/modules/70-pipeline/utils/statusHelpers.ts]: agreed, removed

File: src/modules/10-common/hooks/useDocumentFavicon.ts (hunk 11 above)
Comment (samarth-sharma31) [reply src/modules/10-common/hooks/useDocumentFavicon.ts]: done

File: src/modules/10-common/hooks/useDocumentFavicon.ts (hunk 10 above)
Comment (samarth-sharma31) [reply src/modules/10-common/hooks/useDocumentFavicon.ts]: done

File: src/modules/10-common/hooks/useDocumentFavicon.ts (hunk 6 above)
Comment (samarth-sharma31) [reply src/modules/10-common/hooks/useDocumentFavicon.ts]: resolved

File: src/modules/10-common/hooks/useDocumentFavicon.ts (hunk 8 above)
Comment (samarth-sharma31) [reply src/modules/10-common/hooks/useDocumentFavicon.ts]: done

File: src/modules/10-common/hooks/useDocumentFavicon.ts (hunk 9 above)
Comment (samarth-sharma31) [reply src/modules/10-common/hooks/useDocumentFavicon.ts]: done

**Review by abhinavrastogi-harness — APPROVED** [review]

**Review by kapilkumawat86 — APPROVED** [review]

File: package.json [hunk 13]
```diff
@@ -48,7 +48,7 @@
     "generate-eslint": "node scripts/generate-eslint-config-for-imports.mjs",
     "build-docker": "sh scripts/docker-build.sh",
     "lighthouse": "node scripts/lighthouse/lighthouse.js",
-    "custom-coverage": "npx new-files-coverage-check --e !**/*.test.tsx !**/*.test.ts !**/*.snap !**/*.js !**/*.scss !**/*.mock.ts !**/*.spec.ts !**/*.d.ts !**/*.json !**/*.types.ts !**/types.ts !**/*.yml !**/*.yaml !**cypress/** !**/cypress/** !**/__test*/** !**/*.png !**/*.svg !**/src/services/** !**/src/queries/** !**/src/**/__mocks__/** !**/src/**/RouteDestinations.tsx !**/*.sh !**/*.md !**/index.ts !**/src/modules/75-chaos/**"
+    "custom-coverage": "npx new-files-coverage-check --e !**/*.test.tsx !**/*.test.ts !**/*.snap !**/*.js !**/*.scss !**/*.mock.ts !**/*.spec.ts !**/*.d.ts !**/*.json !**/*.types.ts !**/types.ts !**/*.yml !**/*.yaml !**cypress/** !**/cypress/** !**/__test*/** !**/*.png !**/*.ico !**/*.svg !**/src/services/** !**/src/queries/** !**/src/**/__mocks__/** !**/src/**/RouteDestinations.tsx !**/*.sh !**/*.md !**/index.ts !**/src/modules/75-chaos/**"
```
[the remark this answers is not in the document]
Comment (samarth-sharma31) [reply package.json:51]: sure @abhinavrastogi-harness , i'll do this change in a follow up PR, this one is merged

**Pull request merged**